src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/AggregateCompletionListProvider.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Completion
+{
+    internal class AggregateCompletionListProvider
+    {
+        private readonly IReadOnlyList<CompletionListProvider> _completionListProviders;
+        private readonly ILogger<AggregateCompletionListProvider> _logger;
+
+        public AggregateCompletionListProvider(IEnumerable<CompletionListProvider> completionListProviders, ILoggerFactory loggerFactory)
+        {
+            _completionListProviders = completionListProviders.ToArray();
+            _logger = loggerFactory.CreateLogger<AggregateCompletionListProvider>();
+        }
+
+        public async Task<VSInternalCompletionList?> GetCompletionListAsync(
+            int absoluteIndex,
+            CompletionContext completionContext,
+            DocumentContext documentContext,
+            VSInternalClientCapabilities clientCapabilities,
+            CancellationToken cancellationToken)
+        {
+            var completionListTasks = new List<Task<VSInternalCompletionList?>>(_completionListProviders.Count);
+            foreach (var completionListProvider in _completionListProviders)
+            {
+                try
+                {
+                    var task = completionListProvider.GetCompletionListAsync(absoluteIndex, completionContext, documentContext, clientCapabilities, cancellationToken);
+                    completionListTasks.Add(task);
+                }
+                catch (Exception ex) when (ex is not TaskCanceledException)
+                {
+                    _logger.LogError(ex, "Resolving completions failed synchronously unexpectedly.");
+                }
+            }
+
+            var completionLists = new Queue<VSInternalCompletionList>();
```

Comment:
Is this whole chunk of code not just `var completionLists = (await Task.WhenAll(completionListTasks)).Where(l=>l is not null)`?

Reply:
That doesn't capture error cases 😢